### DISCUSSION TITLE
How to Train a Custom Detection Head?

### DISCUSSION THREAD

**Discussion by mevlt01001** [ultralytics/ultralytics#20747] [Q&A] [answered]
Hello everyone 👋

I'm a computer engineering undergraduate student currently working on a research project at my university.  
My biggest motivation is to learn deeply by experimenting and exploring advanced topics in computer vision.

**My goal:** Combine the backbones of two different object detectors and investigate whether an ensemble model can outperform a single backbone in object detection tasks.

---

## **Feature Maps in My Setup**

After extracting features, I have **three different scales**:
- `f1: [1, C, 80, 80]`
- `f2: [1, C, 40, 40]`
- `f3: [1, C, 20, 20]`

These are fed into a **custom detection head** (not YOLO's standard `Detect` module).

---

## **What I've Understood So Far**
- Each feature map represents a logical grid over the input image.
- E.g., for a 640×640 input, `f1` is an 80×80 grid (so each cell covers an 8×8 region), `f2` is 16×16, and `f3` is 32×32.
- But I couldn't find a clear resource explaining **how training proceeds using these features, especially for anchor-free detection** (no anchors).

---

## **My Core Questions**

1. ### **How does each grid cell “see” a ground-truth box during training?**
    - Are grid cells matched based on the center of the GT box?
    - Is there another mechanism for assigning positive/negative samples in anchor-free methods?

2. ### **How are GT boxes encoded for these multi-scale grids?**
    - In anchor-free detection (like FCOS, YOLOv8-AF), how do we establish the correspondence between a grid cell and a GT box?
    - Why do most recent models prefer ltrb (left/top/right/bottom distances) over cxcywh (center/width/height)?  
      What are the advantages?

3. ### **What is the flow from feature map → head output → loss calculation?**
    - How do we build the target tensors (e.g., for DFL loss)?
    - Is there a code snippet or visual explanation for the ground-truth assignment in multi-scale, anchor-free heads?

---

```python
# Example: f1, f2, f3 as multi-scale feature maps
f1 = torch.randn(1, C, 80, 80)
f2 = torch.randn(1, C, 40, 40)
f3 = torch.randn(1, C, 20, 20)
# → fed into custom detection head
```
Here is my head onnx image:
![detector_only_head onnx](https://github.com/user-attachments/assets/9ee37300-742a-4b1d-a70d-9716bddd1e52)

I’m still a beginner in object detection internals, but I’m very eager to learn and open to feedback.

I’d really appreciate any guidance, references, or code snippets that can help clarify the training process for anchor-free, multi-scale detection heads—especially regarding ground-truth assignment and target construction.

If anyone has experience with custom heads, DFL loss, or similar setups, your insights would be extremely valuable for my project and my learning journey.

Thank you all in advance for your time and support! 🙏

**Comment by UltralyticsAssistant** [comment]
👋 Hello @mevlt01001, thank you for sharing your research project and for your detailed questions about custom detection heads and anchor-free detection! 🚀 Your enthusiasm for learning and experimentation in computer vision is fantastic.

If this is a 🐛 Bug Report, please provide a [minimum reproducible example](https://docs.ultralytics.com/help/minimum-reproducible-example/) to help us debug it.

For custom training or architecture modification ❓ questions like yours, please include as much detail as possible, such as code snippets, dataset examples, and training logs. This helps our team and the community provide more targeted guidance. You may also want to review our [Tips for Best Training Results](https://docs.ultralytics.com/guides/model-training-tips/) for practical insights.

For deeper dives, the [Ultralytics Docs](https://docs.ultralytics.com/) contain comprehensive guides on [YOLO Modes](https://docs.ultralytics.com/modes/) and [Tasks](https://docs.ultralytics.com/tasks/), which may help clarify multi-scale, anchor-free detection flows and ground-truth assignments. You can find [Python](https://docs.ultralytics.com/usage/python/) and [CLI](https://docs.ultralytics.com/usage/cli/) usage examples to complement your experiments.

Join the Ultralytics community for real-time help or to discuss advanced topics:
- [Discord](https://discord.com/invite/ultralytics) 🎧 for chat
- [Discourse](https://community.ultralytics.com/) for in-depth Q&A
- [Subreddit](https://reddit.com/r/Ultralytics) to share knowledge and get feedback

## Upgrade

To ensure you’re working with the latest features and fixes, upgrade the `ultralytics` package and all [requirements](https://github.com/ultralytics/ultralytics/blob/main/pyproject.toml) in a [**Python>=3.8**](https://www.python.org/) environment with [**PyTorch>=1.8**](https://pytorch.org/get-started/locally/):

```bash
pip install -U ultralytics
```

## Environments

YOLO can be run in any of these verified environments (with dependencies and GPU support):

- **Notebooks**: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo11"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud** Deep Learning VM ([GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/))
- **Amazon** Deep Learning AMI ([AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/))
- **Docker Image** ([Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/)) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

## Status

<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml/badge.svg" alt="Ultralytics CI"></a>

If this badge is green, all [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule) tests are currently passing—verifying YOLO [Modes](https://docs.ultralytics.com/modes/) and [Tasks](https://docs.ultralytics.com/tasks/) on macOS, Windows, and Ubuntu.

This is an automated response 🤖. An Ultralytics engineer will review your discussion and provide further assistance soon!

**Reply by glenn-jocher** [reply, marked as answer]
Hello @mevlt01001, great questions about anchor-free detection heads! Let me address your specific questions about how anchor-free detection works in models like YOLO11:

1. **Grid cell assignment**: In anchor-free detection, grid cells are typically assigned to ground-truth objects based on whether the center point of the GT box falls within the cell. This is much simpler than anchor-based methods. Cells containing an object center are considered positive samples, while others are negative. Some methods also use a center sampling strategy where cells near the center point may also be considered positive with varying weights.

2. **GT box encoding**: For anchor-free detection in multi-scale grids:
   - Each grid cell predicts distances (ltrb) from the cell center to the object boundaries
   - The `dist2bbox` and `bbox2dist` functions in Ultralytics [TAL module](https://docs.ultralytics.com/reference/utils/tal/) handle this conversion
   - LTRB is preferred over cxcywh because it provides better localization performance and avoids scale normalization issues when training with various sized objects

3. **Feature map to loss flow**:
   - Feature maps are processed through the detection head to predict class probabilities and box coordinates
   - For each grid cell, the head predicts class scores and regresses the distance values
   - During training, grid cells are assigned to GT boxes based on center points
   - Losses (typically a combination of classification loss, box regression loss, and optionally DFL) are calculated between predictions and assigned targets

The Ultralytics detection head implementation in [`head.py`](https://docs.ultralytics.com/reference/nn/modules/head/) provides a good reference for how this is structured in code.

Would you like me to elaborate on any specific part of this process?
